File: code/divide_conquer/inversion_count/inversion_count.java
//  divide conquer | inversion count | Java
//  Part of Cosmos by OpenGenus Foundation

import java.util.Scanner;
public class InversionCount {
    public static int merge(int a[], int p, int q,int r){

        int i = p ,j = q ,k = 0, count = 0;

        int temp[] = new int[r-p+1];

        while(i<q && j<=r){

            if(a[i] < a[j]){
                temp[k++] = a[i++];
            }
            else{
                temp[k++] = a[j++];
                count += (q - i);
            }
        }

        while(i<q){
            temp[k++] = a[i++];
        }

        while(j<=r){
            temp[k++] = a[j++];
        }

        k = 0;

        while(p<=r)
            a[p++] = temp[k++];

        return count;
    }
    public static int mergeSort(int a[],int i, int j){
        int count = 0;

        if(i>=j)
            return 0;

        int mid = (i+j)/2;

        count += mergeSort(a,i,mid);
        count += mergeSort(a,mid+1,j);
        count += merge(a,i,mid+1,j);

        return count;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter n >> ");
        int n = sc.nextInt();

        int a[] = new int[n];

        System.out.print("Enter elements of array >> ");
        for(int i=0;i<n;i++)
            a[i] = sc.nextInt();

        int count = mergeSort(a,0,a.length-1);

        System.out.println("Number of inversions : " + count);

    }
}

File: code/divide_conquer/src/inversion_count/inversion_count.py
"""

divide conquer | inversion count | Python
part of Cosmos by OpenGenus Foundation

"""


def merge(left, right):
    merged_arr = []
    i = 0
    j = 0
    inv_cnt = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged_arr.append(left[i])
            i += 1
        else:
            merged_arr.append(right[j])
            j += 1
            inv_cnt += len(left) - i

    merged_arr += left[i:]
    merged_arr += right[j:]

    return merged_arr, inv_cnt


def merge_sort(
    arr
):  # this function will return a tuple as (sorted_array, inversion_count)
    if len(arr) > 1:
        mid = len(arr) // 2
        left = arr[:mid]
        right = arr[mid:]

        left, left_inv_cnt = merge_sort(left)
        right, right_inv_cnt = merge_sort(right)
        merged, merged_inv_cnt = merge(left, right)
        return merged, (left_inv_cnt + right_inv_cnt + merged_inv_cnt)
    else:
        return arr, 0


arr = [1, 8, 3, 4, 9, 3]
sorted_array, inversion_count = merge_sort(arr)

print("Sorted array:", sorted_array, " and Inversion count = %s" % inversion_count)
